service/storage_service.cc

```diff
@@ -2552,7 +2563,13 @@ future<> storage_service::raft_state_monitor_fiber(raft::server& raft, sharded<d
                     std::bind_front(&storage_service::raft_topology_cmd_handler, this),
                     _tablet_allocator.local(),
                     get_ring_delay()),
-                [] (std::unique_ptr<topology_coordinator>& coordinator) { return coordinator->run(); });
+                [] (std::unique_ptr<topology_coordinator>& coordinator) -> future<> {
+                    try {
+                        co_await coordinator->run();
+                    } catch (...) {
+                        on_fatal_internal_error(slogger, format("raft topology: unhandled exception in topology_coordinator_run: {}", std::current_exception()));
```

Comment:
Shouldn't we initiate a raft leader step down in this case?

Reply:
Good idea indeed. The question is if this should be  done everywhere we assert() or in a generic code that handles the exit because of an assertion.

Reply:
Assertion failure means there's a bug we need to fix, not do a leader stepdown

Reply:
Could you tell me why it would be beneficial to step down in this case? This `catch` block is supposed to be triggered on unhandled exceptions, wouldn't it make more sense to abort ASAP as we encountered a bug?

Reply:
> Assertion failure means there's a bug we need to fix, not do a leader stepdown

You are not wrong, but if a node crashes it would be beneficial for the cluster to find out about it faster. If there is a safe way to do it (safe in a way that it cannot cause harm even if it did not work properly) then why not do it?

Reply:
> Could you tell me why it would be beneficial to step down in this case? This `catch` block is supposed to be triggered on unhandled exceptions, wouldn't it make more sense to abort ASAP as we encountered a bug?

Stepping does is beneficial because it reduce the time the cluster have no leader. But I would not stop this PR from merging because of this.

Reply:
I'd put it into a broader context. We don't know here whether or not it is a bug - all we know is we caught an exception we can't handle. I think there is a set of reasonable cases when the topology coordinator may fail with an uncaught exception: be it a memory issue or an error from the operating system. In this case we can either fail the node gracefully, or just abort it and disrupt cluster availability.

Reply:
I am also Ok with a follow up patch, but looks like it's a call for an error injection test.

Reply:
> We don't know here whether or not it is a bug - all we know is we caught an exception we can't handle. I think there is a set of reasonable cases when the topology coordinator may fail with an uncaught exception

The point of this try..catch is to simulate `noexcept` but for coroutines; `noexcept` immediately aborts if the function throws. Any exception flying out of topology coordinator code indicates a bug in the code, because the topology coordinator fiber is supposed to handle all exceptions internally.

Reply:
>  I am also Ok with a follow up patch, but looks like it's a call for an error injection test.

That would be like asserting that a boolean is true at some point in the code, and then adding an injection just before that assertion that sets the bool to false, to check that the assertion indeed fires.

This try..catch is only there because of a defect of the C++ language, we shouldn't be having this discussion.

Reply:
Maybe I am missing some context, but what does make you think that coordinator->run() can not fail for a good, valid reason that is not abort_requested?